actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -628,39 +673,32 @@ def commit(options)
 
           raise CookieOverflow if options[:value].bytesize > MAX_COOKIE_SIZE
         end
-    end
-
-    # UpgradeLegacyEncryptedCookieJar is used by ActionDispatch::Session::CookieStore
-    # instead of EncryptedCookieJar if secrets.secret_token and secrets.secret_key_base
-    # are both set. It reads legacy cookies signed with the old dummy key generator and
-    # encrypts and re-saves them using the new key generator to provide a smooth upgrade path.
-    class UpgradeLegacyEncryptedCookieJar < EncryptedCookieJar #:nodoc:
-      include VerifyAndUpgradeLegacySignedMessage
-    end
 
-    # UpgradeLegacyHmacAesCbcCookieJar is used by ActionDispatch::Session::CookieStore
-    # to upgrade cookies encrypted with AES-256-CBC with HMAC to AES-256-GCM
-    class UpgradeLegacyHmacAesCbcCookieJar < EncryptedCookieJar
-      def initialize(parent_jar)
-        super
+        def try_legacy_verifier(name, encrypted_message)
+          deserialize(name, @legacy_verifier.verify(encrypted_message)).tap do |deserialize|
+            self[name] = deserialize
+          end
 
-        secret = key_generator.generate_key(request.encrypted_cookie_salt || "")[0, ActiveSupport::MessageEncryptor.key_len]
-        sign_secret = key_generator.generate_key(request.encrypted_signed_cookie_salt || "")
+        rescue ActiveSupport::MessageVerifier::InvalidSignature
+          nil
+        end
 
-        @legacy_encryptor = ActiveSupport::MessageEncryptor.new(secret, sign_secret, cipher: "aes-256-cbc", digest: digest, serializer: ActiveSupport::MessageEncryptor::NullSerializer)
-      end
+        def extract_key(cipher:, raw_key: nil, key_generator: nil, secret: nil, salt: nil, **)
+          return raw_key if raw_key
+          raise ArgumentError, "Missing salt to generate encryption key from" unless salt
 
-      def decrypt_and_verify_legacy_encrypted_message(name, signed_message)
-        deserialize(name, @legacy_encryptor.decrypt_and_verify(signed_message)).tap do |value|
-          self[name] = { value: value }
+          key_generator ||= ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
+          key_generator.generate_key(salt, ActiveSupport::MessageEncryptor.key_len(cipher))
         end
-      rescue ActiveSupport::MessageVerifier::InvalidSignature, ActiveSupport::MessageEncryptor::InvalidMessage
-        nil
-      end
 
-      private
-        def parse(name, signed_message)
-          super || decrypt_and_verify_legacy_encrypted_message(name, signed_message)
+        def extract_signing_key(cipher:, raw_signed_key: nil, key_generator: nil, secret: nil, signed_salt: nil, **)
+          return unless cipher =~ AES_CBC_CIPHER_REGEX
+          return raw_signed_key if raw_signed_key
+
+          raise ArgumentError, "Missing signed_salt to generate sign key from" unless signed_salt
+
+          key_generator ||= ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
+          key_generator.generate_key(signed_salt)
```

Comment:
Think we should pull these key extractions into the `rotate` methods on the verifier and encryptor.

I worked a bit on a different version of the rotations, which also includes a different approach to what happens on legacy rotations. How does this sound?

```ruby
# frozen_string_literal: true

module ActiveSupport
  module Messages
    module Rotator # :nodoc:
      def initialize(*)
        super

        @rotations = []
        @on_legacy_rotation = nil
      end

      def on_legacy_rotation(on_rotation)
        old_on_legacy_rotation, @on_legacy_rotation = @on_legacy_rotation, on_rotation

        yield
      ensure
        @on_legacy_rotation = old_on_legacy_rotation
      end

      def rotate(*args)
        @rotations << make_rotation(*args)
      end

      private
        def key_generator_for(secret)
          ActiveSupport::KeyGenerator.new(secret, iterations: 1000)
        end

        def legacy_rotation(message)
          @on_legacy_rotation.call(message) if @on_legacy_rotation
        end

        def in_rotations
          @rotations.find do |rotation|
            mark_legacy_rotation(yield(rotation)) rescue next
          end
        end

      module Encryptor
        include Rotator

        def decrypt_and_verify(*args)
          super
        rescue MessageEncryptor::InvalidMessage, MessageVerifier::InvalidSignature => error
          in_rotations { |encryptor| encryptor.decrypt_and_verify(*args) } || raise(error)
        end

        private
          def make_rotation(raw_key: nil, raw_signed_key: nil, **options)
            self.class.new \
              raw_key || extract_key(options),
              raw_signed_key || extract_signing_key(options),
              options.slice(:cipher, :digest, :serializer)
          end
        
          def extract_key(salt:, key_generator: nil, secret: nil, cipher: nil, **)
            key_generator ||= key_generator_for(secret)
            key_generator.generate_key(salt, self.class.key_len(cipher))
          end

          def extract_signing_key(key_generator: nil, signed_salt: nil, secret: nil, **)
            unless cipher =~ /\Aaes-\d+-cbc\z/i
              raise ArgumentError, "missing signed_salt to generate signing key from"
            end

            key_generator ||= key_generator_for(secret)
            key_generator.generate_key(signed_salt)
          end
      end

      module Verifier
        include Rotator

        def verify(*args, &block)
          super
        rescue MessageVerifier::InvalidSignature => error
          in_rotations { |verifier| verifier.verify(*args) } || raise(error)
        end
      end

      private
        def make_rotation(raw_key: nil, digest: nil, serializer: nil, **options)
          self.class.new(raw_key || extract_key(options), digest: digest, serializer: serializer))
        end
      
        def extract_key(key_generator: nil, secret: nil, salt:)
          key_generator ||= key_generator_for(secret)
          key_generator.generate_key(salt)
        end
    end
  end
end
```

Reply:
I agree we should move the key and options extraction to within the rotator. It's more useful that way.

Some questions about this code:
1. Where is `mark_legacy_rotation` defined? Should this be a call to `legacy_rotation` instead?
2. Using `on_legacy_rotation` to signal to the cookies middleware that a rotated instance was using may lead to deserializing the cookie twice. This is an issue I ran into earlier with a similar, callback-based approach.